src/mps/mpo.jl

```diff
@@ -491,14 +492,24 @@ function _contract_densitymatrix(A::MPO, ψ::MPS; kwargs...)::MPS
   R = R * dag(Ut) * ψ[n - 1] * A[n - 1]
   simR_c = simR_c * U * ψ_c[n - 1] * simA_c[n - 1]
   for j in reverse(2:(n - 1))
+    if maxdim_defined
+      maxdim = max_maxdim
+    else
+      maxdim = 1
+      cip = commonind(ψ[j], E[j - 1])
+      ciA = commonind(A[j], E[j - 1])
+      (cip != nothing) && (maxdim *= dim(cip))
+      (ciA != nothing) && (maxdim *= dim(ciA))
+      (cip == nothing && ciA == nothing) && (maxdim = max_maxdim)
```

Comment:
Couldn't we always compute this "locally maximum dimension", and than define maxdim as `min(maxdim, max_maxdim)`?

Also if you use:
```julia
      cip = commoninds(ψ[j], E[j - 1])
      ciA = commoninds(A[j], E[j - 1])
```
then you should just be able to generically use `dim(cip)` and `dim(ciA)`, which are defined to be `1` if there are no indices. This should make the code logic simpler.

Finally, in the case when `cip == nothing && ciA == nothing`, can we safely make `maxdim = 1`? That's what would happen with my suggestion above.

Reply:
I like all those suggestions - thanks. Let me see what I can do to simplify that code, as I agree it's kind of unwieldy.